### PR TITLE
Implement `SharedArrayBuffer`

### PR DESCRIPTION
**Pull request by jedel1043**
I'll follow up this PR with some others implementing Atomics, adding the missing example/example docs and adding the missing `agent` object.

### CODE REVIEW

**Review by raskad — COMMENTED** [review]
Really nice work. I really like the refactor of the typed array kinds!

I still have to go through some of the unsafe call sites to review the invariants.

For others reviewing; it makes sense to compare the old `boa_engine/src/builtins/typed_array/mod.rs` to the new `boa_engine/src/builtins/typed_array/builtin.rs` since most of the contents are moved.

File: boa_engine/src/builtins/array_buffer/shared.rs [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            length: new_len,
+        } = get_slice_range(
+            buf.len() as u64,
+            args.get_or_undefined(0),
+            args.get_or_undefined(1),
+            context,
+        )?;
+
+        // 14. Let ctor be ? SpeciesConstructor(O, %SharedArrayBuffer%).
+
```
Comment (HalidOdat) [inline boa_engine/src/builtins/array_buffer/shared.rs:198]:
```suggestion
```

File: boa_engine/src/builtins/array_buffer/shared.rs [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            // 22. Perform CopyDataBlockBytes(toBuf, 0, fromBuf, first, newLen).
+
+            let first = first as usize;
+            let new_len = new_len as usize;
+
+            // SAFETY: `get_slice_range` will always return indices that are in-bounds.
+            // This also means that the newly created buffer will have at least `new_len` elements
+            // to write to.
+            unsafe { copy_shared_to_shared(&from_buf[first..], to_buf, new_len) }
```
Comment (HalidOdat) [inline boa_engine/src/builtins/array_buffer/shared.rs:240]:
```suggestion
            // 20. Let fromBuf be O.[[ArrayBufferData]].
            let from_buf = buf.data();

            // 21. Let toBuf be new.[[ArrayBufferData]].
            let to_buf = new_buf.data();

            // 22. Perform CopyDataBlockBytes(toBuf, 0, fromBuf, first, newLen).
            let first = first as usize;
            let new_len = new_len as usize;

            // SAFETY: `get_slice_range` will always return indices that are in-bounds.
            // This also means that the newly created buffer will have at least `new_len` elements
            // to write to.
            unsafe { copy_shared_to_shared(&from_buf[first..], to_buf, new_len) }
```

**Review by HalidOdat — APPROVED** [review]
Really nice work!

just some minor nitpicks that don't block merging :)

**Review by raskad — APPROVED** [review]

**Pull request merged**